File: chat_server.cc
#include <deque>
#include <iostream>
#include <list>
#include <memory>

#include <grpc/grpc.h>
#include <grpc++/server.h>
#include <grpc++/server_builder.h>
#include <grpc++/server_context.h>
#include <grpc++/security/server_credentials.h>
#include "chat.grpc.pb.h"

class ChatImpl final : public chat::Chat::Service {
public:
  grpc::Status PostMessage(grpc::ServerContext *context,
                           const chat::PostMessageRequest *request,
                           chat::PostMessageResponse *response) override {
    std::cout << "PM: " << request->user() << ">" << request->text() << std::endl;
    auto message = std::make_shared<chat::Message>();
    message->set_user(request->user());
    message->set_text(request->text());

    response->mutable_message()->CopyFrom(*message);

    PushMessage(std::move(message));

    return grpc::Status::OK;
  }

  grpc::Status StartRTM(grpc::ServerContext *context,
                        const chat::StartRTMRequest *,
                        grpc::ServerWriter<chat::Message> *stream) override {
    std::unique_lock<std::mutex> lock(mutex_);
    auto message_queue = rtm_queues_.emplace(rtm_queues_.end());
    for (;;) {
      cond_.wait(lock, [=] { return !message_queue->empty(); });
      while (!message_queue->empty()) {
        auto message = std::move(message_queue->front());
        message_queue->pop_front();
        lock.unlock();
        const bool ok = stream->Write(*message);
        lock.lock();
        if (!ok) {
          rtm_queues_.erase(message_queue);
          return grpc::Status::OK;
        }
      }
    }
  }

private:
  void PushMessage(std::shared_ptr<const chat::Message> message) {
    {
      std::lock_guard<std::mutex> lock(mutex_);
      for (auto &message_queue : rtm_queues_)
        message_queue.push_back(message);
    }
    cond_.notify_all();
  }

  std::condition_variable cond_;
  std::mutex mutex_;
  std::list<std::deque<std::shared_ptr<const chat::Message>>> rtm_queues_;
};

void RunServer() {
  std::string server_address("0.0.0.0:50051");
  ChatImpl service;

  grpc::ServerBuilder builder;
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  builder.RegisterService(&service);
  std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
  std::cout << "Server listening on " << server_address << std::endl;
  server->Wait();
}

int main(int argc, char** argv) {
  RunServer();
  return 0;
}

File: chat_server.py
#!/usr/bin/env python

"""The Python implementation of the gRPC chat server."""

from threading import Lock
import Queue
import time

import chat_pb2

_ONE_DAY_IN_SECONDS = 60 * 60 * 24

class ChatServicer(chat_pb2.BetaChatServicer):
  """Provides methods that implement functionality of chat server."""

  def __init__(self):
    self.rtm_queues = []
    self.rtm_queues_lock = Lock()

  def stop_rtm(self):
    with self.rtm_queues_lock:
      for message_queue in self.rtm_queues:
        message_queue.put(None)

  def PostMessage(self, request, context):
    print "PM: [" + request.user + "]: " + request.text
    message = chat_pb2.Message(user=request.user, text=request.text)
    with self.rtm_queues_lock:
      for message_queue in self.rtm_queues:
        message_queue.put(message)
    return chat_pb2.PostMessageResponse(message=message)

  def StartRTM(self, request, context):
    message_queue = Queue.Queue()
    with self.rtm_queues_lock:
      self.rtm_queues.append(message_queue)
    while True:
      message = message_queue.get()
      if message is None:
        break
      yield message
    with self.rtm_queues_lock:
      self.rtm_queues.remove(message_queue)

def serve():
  servicer = ChatServicer()

  server_address = '[::]:50051'
  server = chat_pb2.beta_create_Chat_server(servicer)
  server.add_insecure_port(server_address)
  server.start()
  print "Server listening on " + server_address
  try:
    while True:
      time.sleep(_ONE_DAY_IN_SECONDS)
  except KeyboardInterrupt:
    servicer.stop_rtm()
    server.stop(0)

if __name__ == '__main__':
  serve()
